pkg/controller/completed_workflows.go

```diff
@@ -95,26 +97,25 @@ func CompletedWorkflowsSelectorOutsideRetentionPeriod(retentionPeriodHours int,
 		Operator: v1.LabelSelectorOpNotIn,
 		Values:   hoursToKeep,
 	})
-	s.MatchExpressions = append(s.MatchExpressions, v1.LabelSelectorRequirement{
-		Key:      hourOfDayCompletedKey,
-		Operator: v1.LabelSelectorOpIn,
-		Values:   []string{""},
-	})
```

Comment:
sorry for this mistake 
I think all completed workflows will be contained if we remove this match expression.
change operator to v1.LabelSelectorOpDoesNotExist can solve this problem

Reply:
All "old" completed workflows, right? should that be ok given we are trying to fix this existing bug?

Reply:
I think this addresses [this comment](https://github.com/flyteorg/flytepropeller/pull/419/files#r839975491) right? It seems to me that the v1.LabelSelectorOpDoesNotExist solution that @aeioulisa suggested would work. Basically, since we are calling both [the old](https://github.com/flyteorg/flytepropeller/blob/4e4df8a4e520067b2461a42393ddae71e82c52e6/pkg/controller/garbage_collector.go#L136) and [the new](https://github.com/flyteorg/flytepropeller/blob/4e4df8a4e520067b2461a42393ddae71e82c52e6/pkg/controller/garbage_collector.go#L140) garbage collection functions on each iteration, this would make sure that each only processes workflow CRDs that it is responsible for (ie. old function for old labels and new function for new labels).

Reply:
I agree that we should add `v1.LabelSelectorOpDoesNotExist` to avoid garbage collecting things in the wrong iteration.

Reply:
My understanding is that the issue with the existing delete is that it misses out some workflows indefinitely... this would require a one-time delete operation for these workflows... are you all saying this should just go in documentation, @hamersaw, @honnix, @aeioulisa ?

Reply:
Oh OK, I may not have the context for old workflow issues. I don't think it is fair to expect users to manually cleanup old workflows. So this solution will cleanup all old workflows and only recognize the retention period when GCing workflow CRDs moving forward?

Reply:
I feel it might be better we document the manual step to clean up, instead of deleting all old workflows without respecting `hour-of-day`.

Another thing. If we do not check existence of `hour-of-day` together with checking `completed-time`, we might just remove this iteration: https://github.com/flyteorg/flytepropeller/blob/4e4df8a4e520067b2461a42393ddae71e82c52e6/pkg/controller/garbage_collector.go#L136 because everything not having `completed-time` or out of retention period will be cleaned up.

Reply:
@hamersaw, yeah it's not my first choice either... will at least include that in the release notes...
@honnix updated with the additional filter...

Reply:
Nice, thanks!

Reply:
I think the test case needs to be updated.